### PR TITLE
SmartShop Example Update

### PR DESCRIPTION
**Pull request by emmanuelJet**
Fixed hanssv.chain concerns for the SmartShop Sophia Example

### CODE REVIEW

File: examples/SmartShop/contracts/SellerContract.aes [hunk 1]
```diff
@@ -1,35 +1,36 @@
 payable contract Seller =
   record state = 
     { seller_address : address,
+      buyer_address : address,
       status : string,
-      contract_address : address }
+      price : int }
   
-  stateful entrypoint init() : state =
+  entrypoint init(ba : address, price' : int) : state =
```
Comment (mradkov) [inline examples/SmartShop/contracts/SellerContract.aes:8]:
```suggestion
  entrypoint init(buyer_address' : address, price' : int) : state =
```

Comment (mradkov) [inline examples/SmartShop/contracts/SellerContract.aes:8]:
```suggestion
  entrypoint init(buyer_address' : address, price' : int) : state =
```

File: examples/SmartShop/contracts/SellerContract.aes [hunk 2]
```diff
@@ -1,35 +1,36 @@
...
     { seller_address = Call.caller,
+      buyer_address = ba,
```
Comment (mradkov) [inline examples/SmartShop/contracts/SellerContract.aes:10]:
```suggestion
      buyer_address = buyer_address',
```

File: examples/SmartShop/contracts/SellerContract.aes [hunk 3]
```diff
@@ -1,35 +1,36 @@
...
   function check_money() : bool =
-    Contract.balance > 0
+    Contract.balance >= state.price
     
   stateful entrypoint send_item() : string =
     require(check_money(), "Not enough money")
     put(state{status = "sent_to_transport_courier"})
     state.status
+
+  function check_buyer() : bool =
```
Comment (mradkov) [inline examples/SmartShop/contracts/SellerContract.aes:22]: `check_buyer()` is not a very self-explanatory naming for the purpose of this function.

Comment (emmanuelJet) [reply examples/SmartShop/contracts/SellerContract.aes:22]: How about check_state_buyer().

Also for this function in question, the best would have been to check if the `Call.caller == state.buyer_address` but for a reason or the other, this is not executing even after deploying the BuyerContract with another wallet keyPair at the unit test script.

File: examples/SmartShop/contracts/BuyerContract.aes [hunk 4]
```diff
@@ -9,28 +9,20 @@ contract TransportInterface =
   entrypoint check_courier_location : () => string
   
 payable contract Buyer =
```
Comment (hanssv) [inline examples/SmartShop/contracts/BuyerContract.aes:11]: Why did you remove the `state` now you need to pass seller contract and the transport contract in every function?

File: examples/SmartShop/contracts/BuyerContract.aes [hunk 5]
```diff
@@ -1,6 +1,6 @@
 contract SellerInterface =
-  payable stateful entrypoint received_item : () => bool
-  entrypoint get_contract_address : () => address
+  entrypoint received_item : () => unit
+  entrypoint seller_contract_address : () => address
```
Comment (hanssv) [inline examples/SmartShop/contracts/BuyerContract.aes:3]: The function `seller_contract_address` is completely useless...

You would call it like `seller.seller_contract_address()` and then you can just as well just say `seller.address` which is shorter and doesn't involve a remote contract call.

File: examples/SmartShop/contracts/SellerContract.aes [hunk 6]
```diff
@@ -1,35 +1,36 @@
...
+    state.seller_address != state.buyer_address
       
-  payable stateful entrypoint received_item() : bool =
+  stateful entrypoint received_item() : unit =
+    require(check_buyer(), "Buyer Only Function")
```
Comment (hanssv) [inline examples/SmartShop/contracts/SellerContract.aes:26]: So any friend of the seller can call this function? Wouldn't it make more sense if only the buyer could call it?

File: examples/SmartShop/contracts/SellerContract.aes [hunk 7]
```diff
@@ -1,35 +1,36 @@
...
     put(state{status = "delivered" })
     Chain.spend(state.seller_address, Contract.balance)
-    true
 
-  entrypoint key_balance() : int =
-    Chain.balance(Call.caller)
+  entrypoint key_balance() : int = Chain.balance(Call.caller)
```
Comment (hanssv) [inline examples/SmartShop/contracts/SellerContract.aes:30]: A better name would be `caller_balance`?

File: examples/SmartShop/contracts/SellerContract.aes (hunk 3 above)
Comment (emmanuelJet) [reply examples/SmartShop/contracts/SellerContract.aes]: Fixed in the new commit

File: examples/SmartShop/contracts/BuyerContract.aes (hunk 4 above)
Comment (emmanuelJet) [reply examples/SmartShop/contracts/BuyerContract.aes:10]: Fixed in the new commit

File: examples/SmartShop/contracts/BuyerContract.aes (hunk 5 above)
Comment (emmanuelJet) [reply examples/SmartShop/contracts/BuyerContract.aes]: Fixed in the [b182f04 ](https://github.com/aeternity/aepp-sophia-examples/pull/58/commits/b182f04aae62e33f0bdcc7793d954fd996983bd4) commit.

File: examples/SmartShop/contracts/SellerContract.aes (hunk 6 above)
Comment (emmanuelJet) [reply examples/SmartShop/contracts/SellerContract.aes]: Fixed in the [b182f04 ](https://github.com/aeternity/aepp-sophia-examples/pull/58/commits/b182f04aae62e33f0bdcc7793d954fd996983bd4) commit.

File: examples/SmartShop/contracts/SellerContract.aes (hunk 7 above)
Comment (emmanuelJet) [reply examples/SmartShop/contracts/SellerContract.aes]: Fixed in the [b182f04 ](https://github.com/aeternity/aepp-sophia-examples/pull/58/commits/b182f04aae62e33f0bdcc7793d954fd996983bd4) commit.

**Review by hanssv — APPROVED** [review]
Looks much better, good job!
